cdap/resource_local_artifact.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cdap
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func resourceLocalArtifact() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceLocalArtifactCreate,
+		Read:   resourceArtifactRead,
+		Delete: resourceArtifactDelete,
+		Exists: resourceArtifactExists,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"namespace": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				DefaultFunc: func() (interface{}, error) {
+					return defaultNamespace, nil
+				},
+			},
+			// Technically, we could omit the version in the API call because CDAP will infer the
+			// version from the jar. However, forcing the user to specify the version makes dealing
+			// with the resource easier because other API calls require it.
+			"version": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"jar_binary_path": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"json_configuration_path": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceLocalArtifactCreate(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+	ad, err := initArtifactData(config, d)
+	if err != nil {
+		return err
+	}
+	defer ad.Close()
+	if err := uploadJar(config.client, ad); err != nil {
+		return err
+	}
+	if err := uploadProps(config.client, ad); err != nil {
+		return err
```

Comment:
The API will allow you to upload artifacts multiple times and it will just overwrite the old one. I think it is fine to rely on this behaviour and that's what I'm doing here.

If we really want to not rely on that behaviour we'll have to do more than what you suggest because we really care about the contents of the files changing. In particular, the contents of the config file would cause us to have to delete/upload a new artifact because the API calls require you to specify parents.

Thoughts?

Reply:
Ah, ok. I didn't know you can reupload the JAR, thought it would fail. Looks good in that case, thanks for checking!

Reply:
Can you add a comment as well that JARs can be overwritten, thus we don't need partial state management?

Reply:
Done.